scripts/unicode.py

```diff
@@ -432,13 +436,61 @@ def gen_tests(tests, out):
 
     out.write("];\n")
 
+def my_hash(x, salt, n):
+    # This is hash based on the theory that multiplication is efficient
+    mask_32 = 0xffffffff
+    y = ((x + salt) * 2654435769) & mask_32
+    y ^= (x * 0x31415926) & mask_32
+    return (y * n) >> 32
+
+# Compute minimal perfect hash function, d can be either a dict or list of keys.
+def minimal_perfect_hash(d, singleton_buckets = False):
+    n = len(d)
+    buckets = dict((h, []) for h in range(n))
+    for key in d:
+        h = my_hash(key, 0, n)
+        buckets[h].append(key)
+    bsorted = [(len(buckets[h]), h) for h in range(n)]
+    bsorted.sort(reverse = True)
+    claimed = [False] * n
+    salts = [0] * n
+    keys = [0] * n
+    for (bucket_size, h) in bsorted:
+        if bucket_size == 0:
+            break
+        elif singleton_buckets and bucket_size == 1:
+            for i in range(n):
+                if not claimed[i]:
+                    salts[h] = -(i + 1)
+                    claimed[i] = True
+                    keys[i] = buckets[h][0]
+                    break
+        else:
+            for salt in range(1, 32768):
+                rehashes = [my_hash(key, salt, n) for key in buckets[h]]
+                if all(not claimed[hash] for hash in rehashes):
```

Comment:
Is there a guarantee that we won't have a collision amongst the rehashes? Is it just really unlikely? (I suspect it's the latter but want to confirm)

Reply:
Yes, if it finds a suitable salt that comes with a guarantee the rehash won't have a collision (this is what the `claimed` bool-array keeps track of). On the other hand, it's possible that no salt can be found that satisfies that, but I believe it to be quite a low probability. There's things that can be done to make it more robust. I'll try to add a comment outlining that in case somebody does run into it with a data update.

Reply:
Oh, wait, the `set` check deals with this, I'd forgotten it was there :smile: . To be clear, I was specifically worried about cases where a single run of rehashes has collisions, which `claimed` won't catch since we update it later.

(worth leaving a comment saying that)